In [83]:
import pandas as pd

# Работа с Pandas

1. Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк.

Если dataset не задан изначально, его можно загрузить в переменную типа DataFrame из стороннего файла с помощью функции pd.read_'тип файла'('путь к файлу').

3 метода df.head(), df.tail() и df.sample() возвращают первые, последние и случайные пять строк DataFrame. Это полезно, чтобы в первый раз посмотреть какие данные содержит dataset.

In [84]:
df = pd.read_csv('googleplaystore.csv')
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

2. Отобразите количество строк и столбцов в DataFrame.

В датафрейме 10841 строк и 13 столбцов (причём заголовки столбцов в это количество строк не входят).

In [85]:
df.shape

(10841, 13)

3. Выведете основную информацию о DataFrame методом info(). Проанализируйте есть ли пропуски в столбцах. Напишите вывод о том, какие столбцы содержат пропуски.

Пропуски содержатся в столбцах 
2. Rating (1474 пропуска)   существенно больше, чем в остальных
6. Type (1 пропуск)
8. Content Rating (1 пропуск)
11. Current Ver (8 пропусков)
12. Android Ver (3 пропуска) 

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


4. Удалите или замените пропуски известными вам методами. (Подсказка: `dropna()` или `fillna()`, выбор того или иного метода должен быть аргументирован).

Пропущенные данные в существенном количестве есть только в столбце 'Rating'. Поскольку предварительно нет никаких разумных обоснований для того, чтобы заполнять пропущенные данные, например, средним, я предпочитаю удалить все эти строки (учитывая, что их менее 15% от всего dataset).

In [87]:
df.dropna(inplace = True)

5. Проверьте нет ли пропусков после произведенной обработки.
Пропусков больше нет. 

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9360 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9360 non-null   object 
 1   Category        9360 non-null   object 
 2   Rating          9360 non-null   float64
 3   Reviews         9360 non-null   object 
 4   Size            9360 non-null   object 
 5   Installs        9360 non-null   object 
 6   Type            9360 non-null   object 
 7   Price           9360 non-null   float64
 8   Content Rating  9360 non-null   object 
 9   Genres          9360 non-null   object 
 10  Last Updated    9360 non-null   object 
 11  Current Ver     9360 non-null   object 
 12  Android Ver     9360 non-null   object 
dtypes: float64(2), object(11)
memory usage: 1023.8+ KB


6. Выведите на экран минимум и максимум из столбца `Price`:
Максимум 400 у. е. -- приложение 'I'm Rich - Trump Edition'
Минимум -- очень много бесплатных приложенией (8715)

In [89]:
print(df[df['Price'] == df['Price'].max()].shape)
df[df['Price'] == df['Price'].min()].shape

(1, 13)


(8715, 13)

In [90]:
print(df[df['Price'] == df['Price'].max()])
print(df[df['Price'] == df['Price'].min()])

                           App   Category  Rating Reviews  Size Installs  \
4367  I'm Rich - Trump Edition  LIFESTYLE     3.6     275  7.3M  10,000+   

      Type  Price Content Rating     Genres Last Updated Current Ver  \
4367  Paid  400.0       Everyone  Lifestyle  May 3, 2018       1.0.1   

     Android Ver  
4367  4.1 and up  
                                                     App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836           

7. Выведите на экран медиану и среднее арифметическое столбцов:
 - Rating
 - Reviews

 В столбце Reviews возникла проблема 
 TypeError: Cannot convert ['159' '967' '87510' ... '4' '114' '398307'] to numeric
 смотрим тип данных в info(), а там object.

 Самый топорный способ поменять тип данных -- пробежаться по столбцу и поменять тип данных в каждой ячейке.
 Но наверняка это распространённая проблема и для этого существует встроенный метод -- лезем гуглить.

 Гугл предложил 3 способа -- astype(), to_numeric(), convert_dtypes.
 Попробуем astype()

In [91]:
df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price             float64
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [92]:
df['Reviews'] = df['Reviews'].astype(float)
df.dtypes

App                object
Category           object
Rating            float64
Reviews           float64
Size               object
Installs           object
Type               object
Price             float64
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [93]:
print(f'Медиана столбца "Rating": {df["Rating"].median()}')
print(f'Среднее арифметическое столбца "Rating": {df["Rating"].mean()}')
print(f'Медиана столбца "Reviews": {df["Reviews"].median()}')
print(f'Среднее арифметическое столбца "Reviews": {df["Reviews"].mean()}')

Медиана столбца "Rating": 4.3
Среднее арифметическое столбца "Rating": 4.191837606837606
Медиана столбца "Reviews": 5955.0
Среднее арифметическое столбца "Reviews": 514376.7052350427


8. Выведите на экран все уникальные значения категориального столбца Genres.
Воспользуемся методом .value_counts()
115 уникальных значений

In [94]:
df['Genres'].value_counts()

Genres
Tools                      732
Entertainment              533
Education                  468
Action                     358
Productivity               351
                          ... 
Parenting;Brain Games        1
Card;Brain Games             1
Tools;Education              1
Entertainment;Education      1
Strategy;Creativity          1
Name: count, Length: 115, dtype: int64

9. Сгруппируйте данные по столбцу `Genres` и посчитайте для каждого жанра средний и медианный рейтинг. Представьте результат в виде нового DataFrame,который будет сохранен в переменную `grouped_df`, где одна колонка будет содержать названия жарнов, а две другие буду содержать в себе средний и медианный рейтинги. Напишите краткий вывод.

Используем метод group_by()

In [95]:
grouped_df = df.groupby(by='Genres', as_index=False).agg(rating_mean = ('Rating', 'mean'), rating_median = ('Rating', 'median'))
grouped_df

Genres  rating_mean  rating_median
0                                   Action     4.285475            4.3
1                Action;Action & Adventure     4.311765            4.3
2                                Adventure     4.180822            4.3
3             Adventure;Action & Adventure     4.423077            4.5
4                    Adventure;Brain Games     4.600000            4.6
..                                     ...          ...            ...
110                Video Players & Editors     4.063924            4.2
111     Video Players & Editors;Creativity     4.100000            4.1
112  Video Players & Editors;Music & Video     4.000000            4.0
113                                Weather     4.244000            4.3
114                                   Word     4.410714            4.4

[115 rows x 3 columns]

10. Из получившегося датафрейма выведите на экран только те жанры, чей медианный рейтинг больше 4.5. Напишите краткий вывод.

Из 115 различных жанров высокий медианный рейтинг у довольно небольшого (9) числа жанров.

In [96]:
mask = grouped_df.rating_median > 4.5
grouped_df[mask]


Genres  rating_mean  rating_median
4          Adventure;Brain Games     4.600000            4.6
10       Art & Design;Creativity     4.400000            4.7
17            Board;Pretend Play     4.800000            4.8
33             Comics;Creativity     4.800000            4.8
53      Entertainment;Creativity     4.533333            4.6
62    Health & Fitness;Education     4.700000            4.7
72           Music;Music & Video     4.533333            4.6
85              Puzzle;Education     4.600000            4.6
102  Strategy;Action & Adventure     4.600000            4.6

11. Выведете кол-во дубликатов в основном DataFrame (df) и удалите их 

In [97]:
initial_length = len(df)
df.drop_duplicates(inplace = True)
length_without_duplicates = len(df)

In [98]:
print(f'Количество дубликатов в исходной таблице {initial_length - length_without_duplicates}')

Количество дубликатов в исходной таблице 474


12. Посчитатайте какое количество приложений содержится в каждом жанре (value_counts).В получившемся Series выведите на экран 10 строк с наибольшим количеством приложений отсортированных по убыванию.

In [99]:
df['Genres'].value_counts()

Genres
Tools                          732
Entertainment                  498
Education                      446
Action                         349
Productivity                   334
                              ... 
Casual;Music & Video             1
Board;Pretend Play               1
Health & Fitness;Education       1
Music & Audio;Music & Video      1
Strategy;Creativity              1
Name: count, Length: 115, dtype: int64

In [100]:
print(df['Genres'].value_counts().iloc[:10])
# похоже, что value_counts сам делает сортировку по убыванию. Посмотрел документацию -- это действительно так :)
df['Genres'].value_counts().sort_values(ascending=False).iloc[:10]

Genres
Tools              732
Entertainment      498
Education          446
Action             349
Productivity       334
Finance            317
Personalization    308
Communication      307
Lifestyle          304
Photography        304
Name: count, dtype: int64


Genres
Tools              732
Entertainment      498
Education          446
Action             349
Productivity       334
Finance            317
Personalization    308
Communication      307
Lifestyle          304
Photography        304
Name: count, dtype: int64